### PR TITLE
Namespace, Group and User configuration

### LINKED ISSUE #27: Error:  non-standard resources (without the spec field) are not supported.
Altough the rolebinding object is created properly in the matched namespaces,
the NamespaceConfig object reports a failure status: `non-standard resources (without the spec field) are not
    supported` 

Any ideas?

```
kind: NamespaceConfig                                                                                                                                                                         
metadata:                                                                                                                                                                                                                                                                                                                                      
  finalizers:                                                                                                                                                                                 
  - namespace-config-operator                                                                                                                                                                 
  generation: 1                                                                                                                                                                               
  name: kronos-admins-group-rbac                                                                                                                                                              
  namespace: namespace-configuration-operator                                                                                                                                                                                                                                                                                                
spec:                                                                                                                                                                                         
  resources:                                                                                                                                                                                  
  - apiVersion: authorization.openshift.io/v1                                                                                                                                                 
    groupNames:                                                                                                                                                                               
    - kronos-admins                                                                                                                                                                           
    kind: RoleBinding                                                                                                                                                                         
    metadata:
      name: kronos-admins-group
    roleRef:
      name: cluster-admin
    subjects:
    - kind: Group
      name: kronos-admins
  selector:
    matchLabels:
      kronos-rbac: kronos-admins
status:
  lastUpdate: 2019-11-18T10:58:17Z
  reason: "2 errors occurred:\n\t* non-standard resources (without the spec field)
    are not supported\n\t* non-standard resources (without the spec field) are not
    supported\n\n"
  status: Failure
```

### LINKED ISSUE #29: add the ability to define templated resources
this covers for the following issues #28 #21 #20 
a straw-man design is:
1. go templating language
2. the current namespace is passed as the only parameter to the template
3. the same functions as helm will be available to the template engine.

### LINKED ISSUE #33: add the ability to manage users and groups
To complete support for auto-configuration of tenants, we need the ability to potentially create resources when groups or user are created.
Groups are usually create/modified during a group sycn (for example LDAP group sync)
User are created when they login.

Example use cases:
In a situation where a group identifies a dev team working on a project, when a group is created, the namespaces needed to support the project SLDC are create. Also a multi-project resource quota is created.
In a situation where we want to give sandboxes to anyone with a login in OpenShift, when a user is created we also create a namespace and a role for that user in that namespace. The namespace configuration can then assign quota to that namespace.

### LINKED ISSUE #34: Memory leak
Seems the namespace-configuration-operator keeps using more and more memory. After a node suffered some oom issues the log showed a namespace-configuration-operation process being oom-killed by the kernel, releasing 150gb of memory.

As the chart currently does not support setting resource limits/requests for the operator process, I used a k8s `LimitRange` resource to limit the operator to 2gb memory. The following graph shows the operator using more memory as time passes by, until it reaches 2gb and it gets terminated.

namespace-configuration-operator_memory-usage

The cpu resource limit is 2 and this is what docker stats is reporting (it is burning cpu cycles: 204.75%):
```
CONTAINER ID        NAME                                                                                                                                                             CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
f4a81a066fb9        k8s_namespace-configuration-operator_namespace-configuration-operator-68c8b55c4c-s2sg5_namespace-configuration-operator_b8d3e76b-5e9f-492f-9459-ebd08737e72c_0   204.75%             1.977GiB / 2GiB     98.84%              0B / 0B             807GB / 943kB       23
```

### PR DESCRIPTION
**Pull request by raffaelespazzoli**
this PR heavily refactor the namespace-configuration-operator giving the ability to react not only to namespaces but also to user and groups to create arbitrary objects. 
The internal engine has been rewritten and should be much more efficient as it allocate one manager for all the object being  create/watched, as opposed to one manager per object.
Also this refactoring introduces the ability to template the resources.
Finally a generalized object at handling resources has been adopted, so any type of object can be created, not just those who comply with the standard format.

Fixes #34 
Fixes #33 
Fixes #29 
Fixes #27 

I'm opening this PR now, so people interested in those fixes can take a look and comment.
@redmikhail
@jtudelag
@sboschman 
@rh-tstockwell

### CODE REVIEW

File: pkg/controller/groupconfig/groupconfig_controller.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+			log.Error(err, "unable to update instance", "instance", instance)
+			return r.ManageError(instance, err)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	//get selected users
+	selectedGroups, err := r.getSelectedGroups(instance)
+	if err != nil {
+		log.Error(err, "unable to get users selected by", "UserConfig", instance)
```
Comment (sboschman) [inline pkg/controller/groupconfig/groupconfig_controller.go:186]: copy/paste msg from the usersconfig_controller I assume

File: pkg/controller/namespaceconfig/namespaceconfig_controller.go [hunk 2]
```diff
@@ -145,408 +145,164 @@ func (r *ReconcileNamespaceConfig) Reconcile(request reconcile.Request) (reconci
...
 
-func (r *ReconcileNamespaceConfig) getSelectedNamespaces(namespaceconfig *redhatcopv1alpha1.NamespaceConfig) ([]corev1.Namespace, error) {
-	nl := corev1.NamespaceList{}
-	selector, err := metav1.LabelSelectorAsSelector(&namespaceconfig.Spec.Selector)
+	//get selected users
+	selectedNamespaces, err := r.getSelectedNamespaces(instance)
 	if err != nil {
-		log.Error(err, "unable to create selector from label selector", "selector", &namespaceconfig.Spec.Selector)
-		return []corev1.Namespace{}, err
+		log.Error(err, "unable to get users selected by", "UserConfig", instance)
```
Comment (sboschman) [inline pkg/controller/namespaceconfig/namespaceconfig_controller.go:178]: copy/paste from the userconfig_controller I assume

File: pkg/apis/redhatcop/v1alpha1/namespaceconfig_types.go [hunk 3]
```diff
@@ -1,42 +1,54 @@
...
+	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
 
-	Selector  metav1.LabelSelector   `json:"selector,omitempty"`
-	Resources []runtime.RawExtension `json:"resources,omitempty"`
+	// LabelSelector selects Namespaces by label.
+	// +kubebuilder:validation:Required
+	LabelSelector metav1.LabelSelector `json:"labelSelector"`
+
+	// AnnotationSelector selects Namespaces by annotation.
+	// +kubebuilder:validation:Required
```
Comment (sboschman) [inline pkg/apis/redhatcop/v1alpha1/namespaceconfig_types.go:24]: The examples are outdated due to the changed selector:
```
The NamespaceConfig "overcommit-limitrange" is invalid: 
* spec.annotationSelector: Required value
* spec.labelSelector: Required value
```

Do both selectors have to be required though? It feels a bit redundant to add `annotationSelector: {}` to everything if you only want to use the labelSelector.

```
apiVersion: redhatcop.redhat.io/v1alpha1
kind: NamespaceConfig
metadata:
  name: overcommit-limitrange
spec:
  labelSelector:
    matchLabels:
      overcommit: "limited"
  annotationSelector: {}
  resources:    
  - apiVersion: "v1"
    kind: "LimitRange"
    metadata:
      name: "overcommit-limits" 
    spec:
      limits:
        - type: "Container" 
          maxLimitRequestRatio:
            cpu: 100
            memory: 1
```

**Comment by raffaelespazzoli** [comment]
@sboschman I am at the early stages at this point and no image has been build, you are supposed to be testing from your terminal. Those instructions work as long as you have operator-sdk 1.17.
I never found the memory leak, hopefully the new approach will eliminate it, if not it will certainly greatly mitigate it,

File: pkg/apis/redhatcop/v1alpha1/namespaceconfig_types.go (hunk 3 above)
Comment (raffaelespazzoli) [reply pkg/apis/redhatcop/v1alpha1/namespaceconfig_types.go:24]: should have been optional, fixed

**Pull request merged**